src/main/java/io/quarkus/gizmo/Type.java

```diff
@@ -0,0 +1,418 @@
+package io.quarkus.gizmo;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.PrimitiveType.Primitive;
+
+/**
+ * This interface can be used to build a JVM signature for classes, methods and fields.
+ * <p>
+ * Implementations are instantiated via factory methods; for example {@link #voidType()} and {@link #classType(Class)}.
+ * 
+ * @see SignatureBuilder
+ */
+public interface Type {
+
+    // factory methods
+
+    public static VoidType voidType() {
+        return VoidType.INSTANCE;
+    }
+
+    public static PrimitiveType byteType() {
+        return PrimitiveType.BYTE;
+    }
+
+    public static PrimitiveType booleanType() {
+        return PrimitiveType.BOOLEAN;
+    }
+
+    public static PrimitiveType intType() {
+        return PrimitiveType.INT;
+    }
+
+    public static PrimitiveType longType() {
+        return PrimitiveType.LONG;
+    }
+
+    public static PrimitiveType shortType() {
+        return PrimitiveType.SHORT;
+    }
+
+    public static PrimitiveType doubleType() {
+        return PrimitiveType.DOUBLE;
+    }
+
+    public static PrimitiveType floatType() {
+        return PrimitiveType.FLOAT;
+    }
+
+    public static PrimitiveType charType() {
+        return PrimitiveType.CHAR;
+    }
+
+    public static ClassType classType(DotName name) {
+        return new ClassType(Objects.requireNonNull(name).toString().replace('.', '/'));
+    }
+
+    public static ClassType classType(String name) {
+        return new ClassType(name);
```

Comment:
Maybe `requireNonNull` and also `replace('.', '/')`?

Reply:
We have `Objects.requireNonNull()` in the `ClassType` constructor and we don't want to do `replace('.', '/')` because of nested classes; e.g. if you want to use something like `org/acme/Foo<TT;>.Bar`. That's the reason why we take the value as is...

Reply:
I didn't think of nested classes before, but I'd guess an explicit API to express a class/parameterized type that is nested in another class/parameterized type is probably better than having to do this.